src/docker/ids.nim

```diff
@@ -213,72 +227,102 @@ proc isFullyQualified(self: DockerImage): bool =
     maybeRegistry.toLower() != maybeRegistry
   )
 
-proc qualify*(self: DockerImage): DockerImage =
+proc qualify(self: DockerImage): DockerImage =
   ## fully qualify image name with the full registry domain
   ## note qualified name does not specify any scheme like http or https
   ## it simply specifies complete image reference in the registry
   if self.isFullyQualified():
     return self
-  let repo =
-    if '/' in self.repo:
-      self.repo
-    else:
-      "library/" & self.repo
   result = (
-    DEFAULT_REGISTRY & "/" & repo,
+    DEFAULT_REGISTRY & "/" & self.repo,
     self.tag,
     self.digest,
   )
 
-proc normalize*(self: DockerImage): DockerImage =
+proc normalize(self: DockerImage): DockerImage =
   ## normalize qualified registry domain
   ## normalization maps some hardcoded registry domains
   ## to their standard API domains
   # https://github.com/docker/cli/issues/3793#issuecomment-1269051403
-  const registryMapping = {
-    "docker.io":       DEFAULT_REGISTRY,
-    "index.docker.io": DEFAULT_REGISTRY,
-  }.toTable()
-  let qualified = self.qualify()
-  # parseUri doesnt parse uri without any scheme
-  var uri       = parseUri("https://" & qualified.repo)
-  uri.hostname  = registryMapping.getOrDefault(uri.hostname, uri.hostname)
-  let repo = ($uri).removePrefix("https://")
+  let
+    qualified        = self.qualify()
+    (registry, path) = qualified.repo.splitBy("/")
+    fullRegistry     = registry.normalizeRegistry()
+    fullPath         =
+      # library/ is only relevant to docker hub
+      # all external registries are allowed top-level repos
+      if fullRegistry != DEFAULT_REGISTRY or '/' in path:
+        path
+      else:
+        "library/" & path
   result = (
-    repo,
+    fullRegistry & "/" & fullPath,
     self.tag,
     self.digest,
   )
 
-proc uri*(self: DockerImage, scheme = "", path = "", healthcheck = false): Uri =
+proc uri*(self:   DockerImage,
+          scheme  = "",
+          path    = "",
+          prefix  = "",
+          project = "",
+          ): Uri =
   ## generate working URI for the registry API
   ## note this only supports v2 registries hence hardcodes v2 suffix
   ## also this doesnt account for any insecure registry configs
   let normalized = self.normalize()
-  var uri = parseUri("https://" & normalized.repo)
-  if healthcheck:
-    uri.path = "/v2/"
-  else:
-    uri.path = "/v2" & uri.path
+  var uri        = parseUri("https://" & normalized.repo)
+  let uriPath    = uri.path
```

Comment:
does this end up being a value or ref?

Reply:
```nim
Uri = object
  scheme*, username*, password*: string
  hostname*, port*, path*, query*, anchor*: string
  opaque*: bool
  isIpv6*: bool
```

as `path` is `string`, not `ref string`, it will be value